File: leetcode/C++/Word_search.cpp
class Solution {
	public:
		bool exist(vector<vector<char> > &board, string word) {
			if((!word.size()) || (!board.size()))
				return false;
			for(int i = 0; i < board.size(); i++)
			{
				for(int j = 0; j < board[0].size(); j++)
					if(searchword(board, i, j, word, 0))
						return true;
			}
			return false;
		}

		//dfs
		bool  searchword(vector<vector<char> > &board, int row, int col, string &word, int index)
		{
			if(board[row][col] == word[index])
			{
				if(index == (word.size()-1))
					return true;
				char temp = board[row][col];
				board[row][col] = '.';

				//不要单独判定条件，那样就不是dfs，会TLE   
				if(((row > 0) && (searchword(board, row-1, col, word, index+1)))
						||((row < (board.size()-1))&&(searchword(board, row+1, col, word, index+1)))
						||((col > 0) && (searchword(board, row, col-1, word, index+1)))
						||((col < (board[0].size()-1)) && (searchword(board, row, col+1, word, index+1))))
					return true;

				board[row][col] = temp;
			}

			return false;

		}
};

File: leetcode/DFS/Word Search.go
func exist(board [][]byte, word string) bool {
    flags := make([][]bool, len(board))
    for i := 0; i < len(board); i++{
        flags[i] = make([]bool, len(board[0]))
    }

    for i := 0; i < len(board); i++{
        for j := 0; j < len(board[0]); j++{
            contain := dfs(board, i, j, 0, word, flags)
            if contain{
                return true
            }
        }
    }
    return false
}


func dfs(board [][]byte, i int, j int, index int, word string,flags [][]bool) bool{
    if index >= len(word){
        return true
    }
    if i < 0 ||i >= len(board) || j < 0 || j >= len(board[0]) || word[index] != board[i][j] || flags[i][j] == true{
        return false
    }
    flags[i][j] = true
    isContain := dfs(board, i+1, j, index+1, word, flags) ||
    dfs(board, i-1, j, index+1, word, flags) ||
    dfs(board, i, j+1, index+1, word,flags) ||
    dfs(board, i, j-1, index+1, word, flags)
    flags[i][j] = false
    return isContain
}